docs: add contribution section, revise explanations (#6085)
snapcore/snapcraft: docs/contribute/development.rst

Always start by syncing against the branch and the dependencies you're basing your
changes on:

.. tab-set::

    .. tab-item:: Current releases

        If you're contributing to a current release, run:

        .. code-block:: bash

            git fetch upstream
            git switch -c <new-branch-name> upstream/hotfix/<current-release>
            make setup

    .. tab-item:: Future releases

        If you're contributing to multiple releases or the next major release, run:

        .. code-block:: bash

            git switch main
            git pull upstream main
            git switch -c <new-branch-name>
            make setup

Format your branch name as ``<ticket-id>-<description>``. For example, if you're working
on GitHub issue #235, and it's about adding a string sanitizer, you'd name your branch
``issue-235-add-string-sanitizer``.

Develop
~~~~~~~

Snapcraft has conventions for its code style and testing. They are detailed in the
related links on this page. If you're looking for a way to develop something that isn't
covered by a convention, explore the existing codebase.

The Snapcraft codebase is large. You might be tempted to tweak related behavior or fix

> tangential inconsistencies. Don't. Stay focused on your task and keep its scope narrow.

Reviewer: Perhaps add "within a single PR" since we're obviously open to multiple fixes from a single contributor and the paragraph doesn't explicitly state that a PR should be scoped to a single task.

.. meta::
    :description: How to make a code change in Snapcraft. Code, test, and review changes in the project within the standard workflow.

:relatedlinks: [Pytest&#32;test&#32;pattern](https://docs.pytest.org/en/stable/explanation/anatomy.html), [Conventional&#32;Commits](https://www.conventionalcommits.org/en/v1.0.0/), [Spread](https://github.com/canonical/spread/blob/master/README.md)

.. _contribute-development:

Contribute to development
=========================

> This guide describes how to contribute a code change in Snapcraft.

Reviewer 2: Nit:
```suggestion
This guide describes how to contribute a code change to Snapcraft.
```

it on GitHub. Give the PR a title using the same message format as the commit. If your
branch has more than one commit, reuse the message from the first.

If you used an LLM to generate code, explain which parts are synthetic in the PR
description.

Address quality concerns
~~~~~~~~~~~~~~~~~~~~~~~~

Before the PR is merged, it must pass all automatic checks, and it needs separate
approvals from two maintainers.

If there are any issues in your branch that your local testing didn't catch, then the
automatic checks will fail. To address these issues, review the logs in the failed
checks. The error messages in the logs will have remedies and hints for what needs
fixing.

When the maintainers review the PR, they may suggest improvements to your code. Address
them in follow-up commits to your branch, the same way you committed and pushed changes
while drafting. If you feel a particular point should go in a different direction than
what they suggest, discuss it with the maintainer in the PR. They'll be happy to explore
alternatives.

Wrap up the review
~~~~~~~~~~~~~~~~~~

Once all suggestions are addressed, both maintainers approve the PR, and merge it soon

> after. **After the PR is approved, there may be a delay before merge.** The maintainers

Author: Went a slightly different direction here, too. 6ff13e1

> might need time to coordinate the PR with other development on the project.

Reviewer 2: Nit:
```suggestion
might need time to coordinate the PR with other development taking place.
```

After approval, **don't** force-push to your branch. It's difficult for the maintainers
to see whether any additional changes mixed into the push.

Once the PR is merged, your work is complete.

Get help and support
--------------------

> Open source contribution can be difficult. Even the most experienced writers become

Reviewer 2: ```suggestion
Open source contribution can be difficult. Even the most experienced developers become
```

tangled or have moments of uncertainty.

If you're stuck, or need more information about a task, ask the issue creator, the

> project's technical author, or a maintainer. If you need hands-on help, ask in the

Reviewer 2: > the project's technical author

I'm not sure we're the _best_ resource in these cases. 😅

<https://github.com/canonical/snapcraft/issues>`__. If no issue exists for your task,
`create one <https://github.com/canonical/snapcraft/issues/new/choose>`__ for it.

Tasks come in different sizes and complexity. It's important to choose a task that you
have the capacity to finish. It's also important to check to see if the task is already
planned. You might save time and effort by checking for prior work.

Simple tasks
~~~~~~~~~~~~

For obvious tasks that need no explanation, like fixing a lint failure, volunteer for
its GitHub issue. A maintainer will review your request and assign it to you.

Complex tasks
~~~~~~~~~~~~~

For complex tasks, like a new module or class, propose your solution in its GitHub
issue. Describe an implementation, tests, and documentation. If your plan is feasible, a
Snapcraft maintainer will assign the issue to you.

For very complex tasks, like refactors, contact the team in the `Snapcraft Matrix
channel <https://matrix.to/#/#snapcraft:ubuntu.com>`__.

For tasks that require coordination with the Snap Store or snapd, create a post on the
`forum <https://forum.snapcraft.io>`__, which serves as a central point for all teams
working on snaps.

> Bounties

Author: @bepri Can I get your second opinion?

        .. code-block:: bash

            git switch main
            git pull upstream main
            git switch -c <new-branch-name>
            make setup

Format your branch name as ``<ticket-id>-<description>``. For example, if you're working
on GitHub issue #235, and it's about adding a string sanitizer, you'd name your branch
``issue-235-add-string-sanitizer``.

Develop
~~~~~~~

Snapcraft has conventions for its code style and testing. They are detailed in the
related links on this page. If you're looking for a way to develop something that isn't
covered by a convention, explore the existing codebase.

The Snapcraft codebase is large. You might be tempted to tweak related behavior or fix
tangential inconsistencies. Don't. Stay focused on your task and keep its scope narrow.

Most development tasks involve adding or changing algorithms. The architecture, meaning
the organization of the code and the design of its structures, changes much less often.
If in your work you need to change structures or reorganize the code, consult the team
first.

Snapcraft has 10-year long-term support commitments. All changes must be
backward-compatible unless stated otherwise in the issue.

> All nontrivial changes must accompanied by new or updated tests. The Snapcraft test

Reviewer 3: ```suggestion
All nontrivial changes must be accompanied by new or updated tests. The Snapcraft test
```

suite includes both pytest unit tests and Spread integration tests. When adding unit
tests, follow the arrange-act-assert-cleanup pattern from Pytest. Tests are rarely
perfect on the first try. Additional tests can always be added during the review
process.

Snapcraft is amenable to code generated with LLM assistance. Generating code is a time
saver, but like all code it needs testing and careful review. An LLM doesn't absolve you

> of responsibility -- you are ultimately responsible for the code in your work.

Author: I expect many revisions will come to this, so please consider this a best first effort.

In the medium-term, I think the simplest way to cut through the friction is to add a todo item for LLM disclosure to the PR template.
